### DISCUSSION TITLE
ZXing dll not working when we are Printing receipt in Offline mode

### DISCUSSION THREAD

**Discussion by DanyallAslam** [microsoft/Dynamics365Commerce.Solutions#141] [Q&A] [answered]
We are getting below error while printing receipt in offline mode Please note that it is working fine when we are printing receipt in online mode.

**An exception occurred while executing request. CorrelationId: '7fd05ce9-7fec-4094-9af9-a00b4e17d730'. Ex: 'System.IO.FileLoadException: Could not load file or assembly 'zxing, Version=0.16.7.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60'. Could not find or load a specific file. (0x80131621)
File name: 'zxing, Version=0.16.7.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60'
 ---> System.IO.FileLoadException: Could not load file or assembly 'zxing, Version=0.16.7.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60'.
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)**

FnO product version : 10.0.33 (10.0.1549.63)
Commerce app Version: 9.45.23154.2
RSSU Version: 9.45.23153.3

**Comment by FedyukovDA** [comment]
Hi @DanyallAslam, 
As I mention in the previous thread we need more context about your problem.
1) What receipt you are printing, sale receipt, X or Z report or something else?
2) Any special elements on receipt? QR code, bar code?
3) In which country\region your are trying to do that?
4) Any localization is used? If so, then provide more information about what localization is used to print receipt. If it is your own code then provide the sample. 
5) Where you try to print: WindowsPrinter or OPOS device? 
6) Can you confirm that the same scenario is worked in online mode?
7) Do you reference zxing library from your extension?

**Reply by DanyallAslam** [reply]
@FedyukovDA  
1. Custom receipt format
2.  We would like to print QR code 
3. Pakistan
4. we are not using any localization 
5. We are using WindowsPrinter
6. It is working fine in Online mode
7. yes we did add refernece Zxing library in my extension

**Comment by FedyukovDA** [comment]
@DanyallAslam Thank you for the response. 

As I know QR code printing is not supported out of box. So, you have to implement some custom extensions to be able to print it. 
1) As a quick step I can recommend try to reference Zxing library version 0.16.5.991 (or 0.16.5.0 if first is not available).
https://www.nuget.org/packages/ZXing.Net/0.16.5

2) Check that Zxing library exists in the Store commerce extension folder. Default path: C:\Program Files\Microsoft Dynamics 365\10.0\Store Commerce\Extensions\%YourExtensionName%\CommerceRuntime

Reference these guides: 
https://learn.microsoft.com/en-us/dynamics365/commerce/dev-itpro/store-comm-extension-installer
https://community.dynamics.com/ax/b/axforretail/posts/introducing-sealed-installers

**Comment by DanyallAslam** [comment]
@FedyukovDA 

Retail proxy client offline request failed. RequestUri: crt://offline/SalesOrderManager/GetReceipts?queryResultSettings=SCRUBBED&Id=SCRUBBED&receiptRetrievalCriteria=SCRUBBED&$locale=SCRUBBED. Exception: System.IO.FileLoadException: Could not load file or assembly 'zxing, Version=0.16.6.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60'. Could not find or load a specific file. (0x80131621)
File name: 'zxing, Version=0.16.6.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60'
 ---> System.IO.FileLoadException: Could not load file or assembly 'zxing, Version=0.16.6.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60'.

Dll is there, but still we are getting error on Offline mode, and it is working fine on Online mode

![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/125641838/07b6dd0c-7100-4031-949d-e41d7f7965c2)

**Comment by FedyukovDA** [comment]
@DanyallAslam 
Thank you for the testing. A couple of moments from my side:

1. According to the exception, now you reference 0.16.6.0 version of ZXing instead of **0.16.5.0** which I asked to try. 
2. Please check that your extension is loaded to store commerce. To do so follow the next steps:
  a) Install Store Commerce with enabled devtool. You need to set "--enablewebviewdevtools" parameter during installation More information about available parameters here https://learn.microsoft.com/en-us/dynamics365/commerce/dev-itpro/store-commerce#store-commerce-installation-parameters
  b) On a login page open Developers tool (F12) and go the Network tab
  c) Refresh the page (F5) and find "GetExtensionPackageDefinitions()" request
  d) Open "Preview" tab and check the values. There are should be "Contoso.GasStationSample" record. See screenshot example below and please send the same with your data.

![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/124745578/8491cae7-47e1-417f-be62-4faa90231f67)

**Reply by awkaft** [reply]
hi @FedyukovDA ,
![WhatsApp Image 2023-06-16 at 06 29 19](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/126446269/6b818480-88f0-4dfd-8c6b-b3a30d07c662)

**Reply by awkaft** [reply]
Hi @FedyukovDA ,
they preview no nothing.

**Reply by FedyukovDA** [reply]
That's because of cache, let's clean that before.
1) Open dev tool
2) Go to Application tab
3) On the left panel IndexDB -> Microsoft.Dynamics.Commerce.StoreCommerceWeb* -> Cache
4) Then press "Clear object store" icon on the top
5) After that refresh the page and preview should appear.

![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/124745578/b9b66aea-8fdc-4ca0-89dd-eaf2b4cc793d)

**Comment by awkaft** [comment, marked as answer]
Hi @FedyukovDA
please review following Value are visible.
![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/126446269/4ec72d60-587d-4e79-ba51-0f33650196d8)
